etcdctl/ctlv3: auth: wildcard terminated paths specify ranges

glycerine opened the pull request:
Still needs tests, but creating pull request to facilitate the discussion in #6359

xiang90 commented:
What if other people ask for /*/abc? Or /*/abc/*/cef? Should we support these as well? I feel people would ask for globbing eventually, which we do not really want to support in etcd core.

glycerine commented:
> What if other people ask for //abc? Or //abc/*/cef? Should we support these as well? I feel people would ask for globbing eventually, which we do not really want to support in etcd core.

I don't need them or find them compelling.

But maybe I don't understand what the situations they would be used to model.  Could you elaborate on the first one in particular -- what does having two slashes as the first two runes trying to do?

I think this simple PR is perfectly defensible. nats.io and many other pub/sub systems have used similar (filesystem) semantics.

xiang90 commented:
@glycerine Sorry. Github messed up the format. I updated my previous reply.

glycerine commented:
My answer stays the same. : ) 
I don't find them compelling, nor I don't need them. It seems we agree.

glycerine commented:
My implementation is flawed in some way, as tests would have shown.

Where is not obvious... gets seem to work, but put doestn't. When I track it down, it is because the `tocheck` slice in the `checkKeyPerm()` function has length zero during the write attempt; here: https://github.com/coreos/etcd/blob/master/auth/range_perm_cache.go#L148

 Hints appreciated.

glycerine commented:
Yes, I'm mightily confused. Still trying to debug why I have no writes going through.

I have a role, `jason`, that has write permissions.
```
$ etcdctl --user root:123 role get jason
Role jason
KV Read:
        /home/jason/*
KV Write:
        /home/jason/*
```

And yet: inside isRangeOpPermitted(), even if I force a call to getMergedPerms() every time, I still get back no write permissions (at https://github.com/coreos/etcd/blob/master/auth/range_perm_cache.go#L168; with the `if !ok` changed to an `if true`)
```
perms back: '&auth.unifiedRangePermissions{readPerms:[]*auth.rangePerm{(*auth.rangePerm)(0xc42041a0c0)}, writePerms:[]*auth.rangePerm(nil)}'
```

glycerine commented:
sigh, gets (reads) on a "directory" don't work right unless `/home/xiang/*` gets turned into a range.

So just how does one construct the range [0x0, 0xffffffff] ?

glycerine commented:
arg, had to track down of number of insidious bugs in the existing range and range-mergin code. those are fixed now, but I still need tests for the new functionality.

glycerine commented:
If somebody could help out with getting a test that does the following (in library/code rather than shell outs to etcdctl of couse) I would very much appreciate it. I'm out of time for a while to figure out how to write these tests, it would be nice get these fixes and additions in place, and I'm sure somebody is more familiar with the testing facilities already available than I am.

test plan:
```
- create a root user, pw 456, a root role, assign the root role to the root user
- create a regular user called joe, with pw 123, a role called 'regular'
- assign the regular role to the user joe
- give the regular role readonly permissions on /home/*
- create a role called joe, assign it to joe, give it readwrite on /home/joe/*  (joe ends up with 2 roles)
etcdctl auth enable
etcdctl --user root:456 put /home/haddy/hello
etcdctl --user joe:123 put /home/joe/bday 040181
etcdctl --user joe:123 put /home/joe/height "6-0"
- verify that joe can read /home/haddy/hello but not delete it or write a key alongside it. (using the regular role).
- verify that joe can read bday and height from home/joe/ using 'get /home/joe/*'
- verify that joe can delete 2 keys with the library equivalent of : 'etcdctl --user joe:123 del '/home/joe/*'
```

heyitsanthony commented on auth/range_perm_cache.go:
```diff
@@ -22,22 +22,53 @@ import (
 	}
 }
 
+func rangeIsPrefix(a *rangePerm) bool {
```
where is this called?

glycerine replied:
It is simply the inverse of expandPrefixToRange(). Not in use at the moment, but likely to be needed in the near future, possibly in the tests that are TBD. By the way, any pointers re the test plan would be appreciated.

heyitsanthony commented on auth/range_perm_cache.go:
```diff
@@ -161,6 +201,43 @@ func checkKeyPerm(cachedPerms *unifiedRangePermissions, key, rangeEnd []byte, pe
+// Returns true if conversion/expansion took place.
+// If removeWild is true, then we will remove the wildcard byte
+// before creating the range.
+func expandPrefixToRange(r *rangePerm, wildcard byte, removeWild bool) bool {
```
so how would I declare a role on the prefix `abc*`? won't this turn it into a role on the prefix `abc`, matching more than intended?

heyitsanthony commented on etcdserver/apply_auth.go:
```diff
@@ -75,10 +75,66 @@ func (aa *authApplierV3) Range(txnID int64, r *pb.RangeRequest) (*pb.RangeRespon
+// AllowWildcardGets enables get '*' to return all keys,
+// and get '/home/user/*' to return all keys with
+// the prefix '/home/user/'.
+func AllowWildcardGets(r *pb.RangeRequest) {
```
this is essentially the same as `AllowWildcardDeletes`... the core logic should be consolidated into a single function

glycerine replied:
Yes, it is unfortunate. I'm open to ideas about how to do that. They take two different structures though. Refactoring the code to use the same structure would... be a big change.

heyitsanthony commented on etcdserver/apply_auth.go:
```diff
@@ -75,10 +75,66 @@ func (aa *authApplierV3) Range(txnID int64, r *pb.RangeRequest) (*pb.RangeRespon
+}
+
 func (aa *authApplierV3) DeleteRange(txnID int64, r *pb.DeleteRangeRequest) (*pb.DeleteRangeResponse, error) {
+	AllowWildcardDeletes(r)
```
So how do I delete the key `*`?

heyitsanthony commented on etcdserver/apply_auth.go:
```diff
@@ -75,10 +75,66 @@ func (aa *authApplierV3) Range(txnID int64, r *pb.RangeRequest) (*pb.RangeRespon
 	if err := aa.as.IsRangePermitted(&aa.authInfo, r.Key, r.RangeEnd); err != nil {
 		return nil, err
 	}
+	AllowWildcardGets(r)
```
How do I get a key named `*`?

heyitsanthony commented:
It looks like this breaks for any key with a `*` suffix. There's no way this can go in the way it's designed since it restricts the values keys may take even though etcd3 should be a clean binary keyspace.

glycerine commented:
@heyitsanthony what would be acceptable for handing the presence of special characters like `*`? Do you have an approach in mind? I'm open to ideas.

We're used to bash/filesystem interfaces that treat certain characters specially, spaces and `*` at the shell prompt for example. To embed them literally in a filename from bash, one would have to escape them with a backslash.

Does escaping a star `*` with a backslash, as in `\*`, seem reasonable?

glycerine commented:
with the latest:

- Added backslash escaping to allow `\*` in the presence of wildcard `*`.
- Addressed the other comments by centralizing wildcard handling into a separate package.
- Addressed the code duplication by defining an interface, BegEndRange, to unify treatment of ranges.

glycerine commented:
Added a test file.

@xiang90, @heyitsanthony -- I think this is in good shape. Please take a look.

heyitsanthony commented:
@glycerine anything that imposes special structure on the keys can't be part of the base KV API. Expecting people to escape `*` needlessly complicates the key semantics and absolutely breaks the contract that etcd3 will accept arbitrary binary keys without modification.

This really should be done with a proxy.

glycerine commented:
conclusion was that this won't be the way to go. Closing this, and opening a separate pr with the bug fixes to range handling.
